### PR TITLE
migrate to eslint & update dependencies

### PR DESCRIPTION
**Pull request by boranx**
Fix: https://github.com/kubernetes/test-infra/issues/25896

execution of the validations:
```shell
$ make verify-eslint
hack/make-rules/verify/eslint.sh
yarn install v1.22.19
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.22s.

.
.
✖ 3168 problems (3024 errors, 144 warnings)
  2690 errors and 0 warnings potentially fixable with the `--fix` option.
```

### CODE REVIEW

**Comment by chaodaiG** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-verify-lint/1552297835116892160, lot of the errors are indentation error, searched "indentation" in the log and found 2568 occurrence, is there a tool to auto correct them?

File: prow/cmd/webhook-server/main.go [hunk 1]
```diff
@@ -35,10 +35,10 @@ import (
 )
 
 const (
-	secretID  = "prowjob-webhook-ca-cert"
-	caCert = "ca-cert"
-	caPrivKey = "ca-priv-key"
-	caBundle = "ca-bundle"
+	secretID       = "prowjob-webhook-ca-cert"
```
[the remark this answers is not in the document]
Comment (boranx) [reply prow/cmd/webhook-server/main.go]: those are probably leftovers from go-lint after I executed `make`, removed :+1:

**Comment by boranx** [comment]
@chaodaiG passing `--fix` flag to `eslint` makes them resolved[1] but it creates diff as they are not committed
I think that should be done local and then needs to be pushed

1- e.g: `node_modules/eslint/bin/eslint -c eslintrc.js --fix --ext .ts .`

**Comment by chaodaiG** [comment]
sounds good, could you push the diff in a new commit so that it's easier to review?

**Comment by boranx** [comment]
ack, the option `--fix` has been included and the diff added.

File: hack/run-in-node-container.sh [hunk 2]
```diff
@@ -22,8 +22,8 @@ set -o pipefail
 REPO_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)"
 cd "${REPO_ROOT}"
 
-# NOTE: bazel is currently on node 12.x, but we should probably upgrade at some point
-NODE_IMAGE='node:12-bullseye-slim'
+# NOTE: bazel is currently on node 14.x, but we should probably upgrade at some point
```
Comment (chaodaiG) [inline hack/run-in-node-container.sh:25]: this comment is irrelevant any more

File: package.json [hunk 3]
```diff
@@ -18,17 +18,21 @@
...
+    "@typescript-eslint/eslint-plugin": "^5.31.0",
+    "@typescript-eslint/parser": "^5.31.0",
+    "eslint": "^8.20.0",
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-prefer-arrow": "^1.2.3",
     "jasmine": "~3.4.0",
     "rollup": "2.3.4",
     "rollup-plugin-amd": "4.0.0",
+    "rollup-plugin-terser": "5.1.3",
     "terser": "4.4.0",
```
Comment (chaodaiG) [inline package.json:35]: I might be wrong but iirc this is what we use, should we also bump it?

File: package.json [hunk 4]
```diff
@@ -18,17 +18,21 @@
...
     "@types/gtag.js": "^0.0.0",
     "@types/jasmine": "~3.3.13",
     "@types/pako": "^1.0.1",
+    "@typescript-eslint/eslint-plugin": "^5.31.0",
+    "@typescript-eslint/parser": "^5.31.0",
+    "eslint": "^8.20.0",
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-prefer-arrow": "^1.2.3",
     "jasmine": "~3.4.0",
     "rollup": "2.3.4",
```
Comment (chaodaiG) [inline package.json:32]: do we also want to bump this version?

**Comment by boranx** [comment]
@chaodaiG if I add the diff, the CI fails: 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-integration/1552343733867909120
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-prow-image-build-test/1552343733909852160
any idea why that might be happening?

btw I've reverted the changes back due to CI issue, wfm on local for `make verify-eslint`

**Comment by chaodaiG** [comment]
> @chaodaiG if I add the diff, the CI fails: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-integration/1552343733867909120 https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/test-infra/26934/pull-test-infra-prow-image-build-test/1552343733909852160 any idea why that might be happening?
> 
> btw I've reverted the changes back due to CI issue, wfm on local for `make verify-eslint`

these tests both compile typescripts as part of the prow images building, the typescript change you made probably didn't work with the typescript toolchain versions that are used in this repo. You'll need to update them according to the comments that I left on your PR. These tools https://github.com/kubernetes/test-infra/blob/922f2b0d911f48fc4c38ce86f4b4aa3d5f6aac83/hack/ts-rollup/main.go#L40-L43 in particular are used in typescript compilation, and their versions are controlled by package.json

File: hack/run-in-node-container.sh (hunk 2 above)
Comment (boranx) [reply hack/run-in-node-container.sh]: ack, removed

File: package.json (hunk 3 above)
Comment (boranx) [reply package.json]: I'm not sure, I've updated this to the latest anyway

File: package.json (hunk 4 above)
Comment (boranx) [reply package.json]: again for this one, I've updated to the latest

**Comment by chaodaiG** [comment]
The typescript compilation in prow integration tests and image build tests are triggered via https://github.com/kubernetes/test-infra/blob/master/prow/cmd/deck/gather-static.sh, as it's invoked from Go program, it's a bit hard to read its output. I think it will be easier to debug the type script compilation failure by directly invoke this script from your terminal:
```
$ ./prow/cmd/deck/gather-static.sh
```

**Comment by boranx** [comment]
hey @chaodaiG 

I think I found the issue, applying the changes from `--fix` causing the integration tests to be failing for the following files:

```git
        modified:   prow/cmd/deck/static/command-help/command-help.ts
        modified:   prow/cmd/deck/static/pr/pr.ts
```
so I've reverted changes on these two files back in the hope to see the pipelines pass

**Comment by boranx** [comment]
yeah so there's only linting job left as failing, that's because there are issues reported by linter:

```shell
✖ 1667 problems (1544 errors, 123 warnings)
  1200 errors and 0 warnings potentially fixable with the `--fix` option.

make: *** [Makefile:68: verify-eslint] Error 1
```

**Comment by chaodaiG** [comment]
> yeah so there's only linting job left as failing, that's because there are issues reported by linter:
> 
> ```shell
> ✖ 1667 problems (1544 errors, 123 warnings)
>   1200 errors and 0 warnings potentially fixable with the `--fix` option.
> 
> make: *** [Makefile:68: verify-eslint] Error 1
> ```

This seems expected to me. Our typescript toolchains versions, as you probably have seen, are pretty ancient, and the goal here is to bring them up-to-date. Even though the issue says update the linter for typescript, I can totally imagine that updated linter would conflict with the rest of the typescript compilation tools because they are too old. And just bump the version of the typescript compilation tools might not be sufficient to fix all of the compilation errors, as there might be some usages/patterns not caught by linters, and unfortunately they might need to be fixed manually, which can be started by running the `gather-statics.sh` script as I mentioned in the previous comment.

**Comment by boranx** [comment]
indeed, but that sounds more like refactoring and needs to be worked on more so it might worth to be issuing this as a new item to work on suggestions on linter/refactoring...

I'm not sure I can cover all reported issues in this PR in a short-term time period as there are 461 problems (343 errors, 118 warnings) to be fixed

**Comment by chaodaiG** [comment]
> indeed, but that sounds more like refactoring and needs to be worked on more so it might worth to be issuing this as a new item to work on suggestions on linter/refactoring...
> 
> I'm not sure I can cover all reported issues in this PR in a short-term time period as there are 461 problems (343 errors, 118 warnings) to be fixed

fair enough! Would you please create a separate issue as a feature request of refactoring typescripts and log all your findings there?

**Comment by boranx** [comment]
sure! https://github.com/kubernetes/test-infra/issues/26959 has been created to track the issues reported by the linter

**Comment by boranx** [comment]
hey @chaodaiG,

I found some time today to look at this PR and fix the issues, I've applied some of the suggestions, and some of them (the most reoccurring ones) transformed to be printed as warn, see the `eslintrc.js`

```
        "@typescript-eslint/require-await": "warn",
        "@typescript-eslint/no-floating-promises": "warn",
        "@typescript-eslint/no-unsafe-call": "warn",
        "@typescript-eslint/no-unsafe-return": "warn",
        "@typescript-eslint/no-unsafe-argument": "warn",
        "@typescript-eslint/no-unsafe-assignment": "warn",
        "@typescript-eslint/no-misused-promises": "warn",
``` 

Eventually, unblocking rules one by one and fixing them might be a way of gracefully making changes instead of a bulk change. Currently, all are being reported as warnings and doesn't break anything on CI as far as I can see:

```shell
✖ 386 problems (0 errors, 386 warnings)
```

**Comment by chaodaiG** [comment]
Understand that it's a big ask to do two things in one PR: update lint + update typescript compilation tools. This PR at least updated to eslint and the node version, so this LGTM. However, I would prefer to separate the generated `eslintrc.js` with your update into two separate PRs to makes it more clear to future readers, namely:

```
        "@typescript-eslint/require-await": "warn",
        "@typescript-eslint/no-floating-promises": "warn",
        "@typescript-eslint/no-unsafe-call": "warn",
        "@typescript-eslint/no-unsafe-return": "warn",
        "@typescript-eslint/no-unsafe-argument": "warn",
        "@typescript-eslint/no-unsafe-assignment": "warn",
        "@typescript-eslint/no-misused-promises": "warn",
```

should be in a separate commit. Could you squash this PR to include two commits? the first one is almost everything except for any manual updates in `eslintrc.js` file

**Comment by boranx** [comment]
ack, understood. I've gathered them in 2 commits, the second one just reflects the `eslint` tweak required for ci

**Comment by chaodaiG** [comment]
awesome! thank you! One more thing, would you be able to run https://github.com/kubernetes/test-infra/tree/master/prow/cmd/deck#running-deck-locally locally to make sure that the UI won't crash?

**Comment by boranx** [comment]
works well on my local: http://localhost:8080/

**Comment by chaodaiG** [comment]
great!
/label tide/merge-method-merge

**Comment by chaodaiG** [comment]
/lgtm
/approve
/hold
let's merge it after the automated bump later today(~4:30PM PST)

**Review by chaodaiG — COMMENTED** [review]
/unhold

**Review by chaodaiG — APPROVED** [review]

**Comment by chaodaiG** [comment]
/remove-label tide/merge-method-squash
this PR was squashed by the author already

**Pull request merged**